RSA PSS scheme in signing tools

lsubash opened the pull request:
Added an build config _SIGNING_SCHEME for
selection of signing schmemes. Updated tools
with param for selecting the signing scheme when
they are run in stand alone mode.

Supported Signing schemes - RSA_PCKS_1_5, RSA_PSS
Intel Crypto recommends PSS and same defaulted in
SlimBoot.

PSS Verification patch https://github.com/slimbootloader/slimbootloader/pull/640

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

mauricema commented on BootloaderCorePkg/Tools/GenCapsuleFirmware.py:
```diff
@@ -480,7 +480,8 @@ def ValidateRegistryFormatGuid(Argument):
     parser.add_argument('-k',  '--priv_key', dest='PrivKey', type=str, required=True, help='Private RSA 2048 key in PEM format to sign image')
-    parser.add_argument('-a',  '--alg_hash', dest='HashType', type=str, default='SHA2_256', help='Hash type for signing. Supported types SHA2_256')
+    parser.add_argument('-a',  '--alg_hash', dest='HashType', type=str, default='SHA2_256', help='Hash type for signing. Supported types SHA2_256, SHA2_384')
+    parser.add_argument('-s',  '--sign_scheme', dest='SignScheme', type=str, default='RSA_PSS', help='Signing Scheme. Supported types RSA_PCKS_1_5, RSA_PSS')
```
can we use  "choice" to limit the options here for alg_hash and sign_scheme ?

mauricema commented on BootloaderCorePkg/Tools/GenCapsuleFirmware.py:
```diff
@@ -537,7 +538,7 @@ def ValidateRegistryFormatGuid(Argument):
     # Create final capsule
     #
-    SignImage(Result, args.NewImage, args.HashType, args.PrivKey)
+    SignImage(Result, args.NewImage, args.HashType, args.SignScheme args.PrivKey)
```
does this pass execution ?      Missing a  "," between  args.SignScheme and args.PrivKey.

mauricema commented on BootloaderCorePkg/Tools/GenContainer.py:
```diff
@@ -714,7 +723,7 @@ def create_container (args):
         container_list[0][0] = tuple(hdr_entry)
 
-    gen_container_bin (container_list, out_dir, comp_dir, key_dir, tool_dir)
+    gen_container_bin (container_list, out_dir, comp_dir, sign_scheme, key_dir, tool_dir)
```
This might not the proper way to handle it.
Container might have multiple components and might have different signing scheme,  we should not use a top level parameter for this.  Instead,  this info needs to be put into the layout text to be used.

lsubash replied:
Adding to layout we would need to enforce all container users to provide this input as part of layout.

#    Name ,  ImageFile      ,  CompAlg  ,  AuthType ,         Sign Schme  KeyFile                           ,  Size
# ====================================================================================================
  ( 'BOOT', 'IASIMAGE.bin'  , 'CLASSIC' , 'RSA2048_SHA2_256',  '**RSA_PSS**'  'TestSigningPrivateKey.pem'       , 0, 0),
  ( 'CMDL', 'cmdline.txt'   , 'Dummy'   , 'RSA2048_SHA2_256',  '**RSA_PSS**'  'TestSigningPrivateKey.pem'       , 0, 0),
  ( 'KRNL', 'vmlinuz'       , 'Lz4'    ,  'RSA2048_SHA2_256',  '**RSA_PSS**'  'TestSigningPrivateKey.pem'       , 0, 0),
  ( 'INRD', 'initrd'        , 'Lzma'    , 'SHA2_256',                      ''                                , 0, 0),

Alternatively could we add another auth type as,
RSA2048_SHA2_256_**PSS**
RSA3072_SHA2_384_**PSS** ?

mauricema replied:
you alternative solution looks good to me.

lsubash replied:
Thanks.  Container would support following auth types,
"NONE"               
"SHA2_256"            
"SHA2_384"             
"RSA2048_SHA2_256"     
"RSA3072_SHA2_384"    
"RSA**PSS**2048_SHA2_256"     
"RSA**PSS**3072_SHA2_384"

myguitar commented on BootloaderCorePkg/Tools/CommonUtility.py:
```diff
@@ -209,26 +209,56 @@ def run_process (arg_list, print_cmd = False, capture_out = False):
+        hashdata = fin.read()
+    fin.close()
+
+    pos = hashdata.index("=")
+    hashdata = hashdata[pos+2:pos+2+64]
```
if '=' is not found?
if the size of hashdata is not 64?
Can you also use try/except?

myguitar commented on BootloaderCorePkg/Tools/CommonUtility.py:
```diff
@@ -209,26 +209,56 @@ def run_process (arg_list, print_cmd = False, capture_out = False):
+    hash_file     = out_file+'.hash'
+
+    # Generate hash using openssl dgst in hex format
+    cmdargs = [get_openssl_path(), 'dgst', '-'+'%s' % _hash_type_string[hash_type], '-out', '%s' % hash_file_tmp, '%s' % in_file]
```
'-'+'%s' -> '-%s'

lsubash commented:
Update following with this version of patch,
- Fixed review comments in GenCapulefirmware.
- Introduced new auth types in Container instead of introducing signing scheme to layout.
  Existing auth types names are modified for PKCS1
            "NONE"                       : 0,
            "SHA2_256"                   : 1,
            "SHA2_384"                   : 2,
            "RSA2048_**PKCS1_**SHA2_256"     : 3,
            "RSA3072_PKCS1_SHA2_384"     : 4,
            "RSA2048_**PSS_**SHA2_256"       : 5,
            "RSA3072_PSS_SHA2_384"       : 6,

Platform board config require update for Container auth types.,

mauricema commented on BootloaderCorePkg/Tools/GenContainer.py:
```diff
@@ -626,15 +644,23 @@ def gen_container_bin (container_list, out_dir, inp_dir, key_dir = '.', tool_dir
     return auth_type_str
 
 def gen_layout (comp_list, img_type, auth_type_str, out_file, key_dir, key_file):
+    print ("auth_type_str1 %s" % auth_type_str)
```
is this for debug only ?

lsubash replied:
Yes its for debug purposes. I would clean it up.

myguitar commented on BootloaderCorePkg/Tools/CommonUtility.py:
```diff
@@ -209,26 +209,60 @@ def run_process (arg_list, print_cmd = False, capture_out = False):
+        hashdata = fin.read()
+    fin.close()
+
+    pos = hashdata.index("=")
+    if hash_type == "SHA2_256":
+        hashdata = hashdata[pos+2:pos+2+64]
+    elif hash_type == "SHA2_384":
+        hashdata = hashdata[pos+2:pos+2+96]
```
Still recommend to use try: except:
what about else? SHA2_512?

lsubash replied:
SHA512 is not supported currently.
Thanks. I will add try:expection error checks.

lsubash replied:
Done.

lsubash commented:
This PR need to merged after PR# 640

myguitar commented on BootloaderCorePkg/Tools/CommonUtility.py:
```diff
@@ -209,26 +209,65 @@ def run_process (arg_list, print_cmd = False, capture_out = False):
+        hashdata = fin.read()
+    fin.close()
+
+    pos = hashdata.find("=")
+    if pos !=  -1:
+        if hash_type == "SHA2_256":
+            hashdata = hashdata[pos+2:pos+2+64]
+        elif hash_type == "SHA2_384":
+            hashdata = hashdata[pos+2:pos+2+96]
+        else:
+            raise Exception('Unknow supported signing hash type "%s" !' % hash_type)
+    else:
+        raise Exception('Hash Data not found for signing!')
```
I'd like to suggest like this.

try:
    hashdata = hashdata.rsplit('=', 1)[1].strip()
    print(len(hashdata))
except:
    print('something wrong')
    raise

lsubash replied:
Thanks Aiden for suggestion.
Updated,

myguitar commented on BootloaderCorePkg/Tools/GenContainer.py:
```diff
@@ -108,19 +110,34 @@ def __init__(self, buf = None):
-    "SHA2_384"             : "SHA2_384",
-    "RSA2048_SHA2_256"     : "SHA2_256",
-    "RSA3072_SHA2_384"     : "SHA2_384",
+    "NONE"                       : "NONE",
+    "SHA2_256"                   : "SHA2_256",
+    "SHA2_384"                   : "SHA2_384",
+    "RSA2048_PKCS1_SHA2_256"     : "SHA2_256",
+    "RSA3072_PKCS1_SHA2_384"     : "SHA2_384",
+    "RSA2048_PSS_SHA2_256"       : "SHA2_256",
+    "RSA3072_PSS_SHA2_384"       : "SHA2_384",
```
Indent here

myguitar commented on BootloaderCorePkg/Tools/GenContainer.py:
```diff
@@ -108,19 +110,34 @@ def __init__(self, buf = None):
+
+
+    _auth_to_signscheme_str = {
+    "NONE"                       : "",
+    "SHA2_256"                   : "",
+    "SHA2_384"                   : "",
+    "RSA2048_PKCS1_SHA2_256"     : "RSA_PKCS1",
+    "RSA3072_PKCS1_SHA2_384"     : "RSA_PKCS1",
+    "RSA2048_PSS_SHA2_256"       : "RSA_PSS",
+    "RSA3072_PSS_SHA2_384"       : "RSA_PSS",
```
Indent here

lsubash replied:
Done

lsubash commented:
Automated Qemu execution test fails due to PR# 640 dependency.

Aiden, Are we doing python2  build for CFL checks?

myguitar commented:
> Automated Qemu execution test fails due to PR# 640 dependency.
> 
> Aiden, Are we doing python2 build for CFL checks?

The final travis result comes from QEMU FWU test. If it's #640 dependency, let's wait until it's merged. Please take a look travis log to make sure.
